### PR TITLE
Improve memory accounting in JOIN and some aggregations

### PR DESCRIPTION
**Pull request by cberner**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java [hunk 1]
```diff
@@ -173,7 +175,45 @@ public void addPage(Page page, int partitionId, Block partitionIds)
...
+    }
+
+    public void compact()
+    {
+        for (int channel = 0; channel < types.size(); channel++) {
+            Type type = types.get(channel);
+            ObjectArrayList<Block> blocks = channels[channel];
+            for (int i = nextBlockToCompact; i < blocks.size(); i++) {
+                Block block = blocks.get(i);
+                if (block.getSizeInBytes() < block.getRetainedSizeInBytes()) {
```
Comment (yaoxin226) [inline presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java]: I am guessing Block.getSizeInBytes() is the compacted memory usage, Block.getRetainedSizeInBytes() is the memory usage with some overhead, like pre-allocated spaces.
So getSizeInBytes() should always less or equal than getRetainedSizeInBytes().

But in function getExpectedCompactedSize(), you are using getRetainedSizeInBytes() to get expected compacted size. Is there a problem here, or I misunderstand these two functions.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/SliceArrayBlock.java [hunk 2]
```diff
@@ -118,4 +126,18 @@ static int getSliceArraySizeInBytes(Slice[] values)
...
+        long sizeInBytes = SizeOf.sizeOf(values);
+        for (Slice value : values) {
+            if (value != null) {
+                sizeInBytes += value.getRetainedSize();
+            }
+        }
+        if (sizeInBytes > Integer.MAX_VALUE) {
+            sizeInBytes = Integer.MAX_VALUE;
+        }
+        return (int) sizeInBytes;
```
Comment (yaoxin226) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/SliceArrayBlock.java]: How about return (int) Math.min(sizeInBytes, Integer.MAX_VALUE);

File: presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java (hunk 1 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java]: Oops, I meant to remove this getExpectedCompactedSize()  before putting it up for review. Good catch! getRetainedSizeInBytes is the amount of heap memory retained by the Block, whereas getSizeInBytes is the logical size of the Block. This is particularly important when you build a Block that has a Slice inside it which has a length()  much smaller than the array it references

**Comment by dain** [comment]
Looks good

**Comment by haozhun** [comment]
I read e1ac346 and have a question about @cberner's change to `SliceArrayBlock.deepCopyAndCompact`. It might be a typo. Should `distinctValues.put(distinct, distinct);` be `distinctValues.put(slice, distinct);` instead? Now, if this is indeed a typo, does it mean this memory optimization is unnecessary?

**Comment by cberner** [comment]
Yep, that's a typo. It was originally a HashMap and I didn't change that call when I changed the type. The optimization is useful to fix. It's there for dictionary encoded ORC blocks

**Pull request merged**